File: src/web/api/decisions.py
"""决策日志 API(B6, 2026-09-18) —— "信号 → 结果"的账, 只读。

两个端点:
- `GET /api/decisions/stats` —— 按信号类型统计 T+1/3/5 命中率;
  **样本不足 (`n < min_sample`) 时不给命中率**, 返回 `insufficient=true`(页面显示"样本不足")。
- `GET /api/decisions/log`    —— 最近的信号明细(含当时价格/上下文快照与回填状态)。

口径: 缺价格 / 未回填一律 NULL(不补 0), 平盘记未命中 —— 本端点只透传, 不在 API 层做任何推算。
"""
from __future__ import annotations

import logging

from fastapi import APIRouter, Depends, Query
from sqlalchemy import text

from src.web.api.auth import get_current_user
from src.web.models import User

logger = logging.getLogger(__name__)

router = APIRouter(tags=["decisions"])


@router.get("/stats")
def decisions_stats(
    days: int = Query(180, ge=7, le=730, description="回看天数(按信号产生日)"),
    min_sample: int = Query(30, ge=1, le=500, description="低于此样本量不给命中率"),
    _: User = Depends(get_current_user),
):
    """按信号类型统计命中率(样本不足则如实说"样本不足", 不拿小样本算百分比)。"""
    from src.core.decision_log import stats
    from src.db.session import get_read_engine

    return stats(get_read_engine(), days=days, min_sample=min_sample)


@router.get("/log")
def decisions_log(
    kind: str | None = Query(None, description="按信号类型过滤, 如 resonance3"),
    limit: int = Query(100, ge=1, le=500),
    _: User = Depends(get_current_user),
):
    """最近的信号明细: 当时价格 + 上下文快照 + T+1/3/5 回填状态(未回填=None)。"""
    from src.db.session import get_read_engine

    sql = (
        "SELECT signal_kind, symbol, trade_date, price_at_signal, context_json, source, "
        "ret_t1, hit_t1, ret_t3, hit_t3, ret_t5, hit_t5, filled_at "
        "FROM decision_log "
        + ("WHERE signal_kind = :kind " if kind else "")
        + "ORDER BY trade_date DESC, id DESC LIMIT :lim"
    )
    params: dict[str, object] = {"lim": int(limit)}
    if kind:
        params["kind"] = kind

    with get_read_engine().connect() as conn:
        rows = conn.execute(text(sql), params).fetchall()

    items = [
        {
            "signal_kind": r[0],
            "symbol": r[1],
            "trade_date": r[2],
            "price_at_signal": None if r[3] is None else float(r[3]),
            "context": r[4] or "",
            "source": r[5] or "",
            "outcomes": {
                label: {
                    "ret": None if r[i] is None else float(r[i]),
                    "hit": None if r[i + 1] is None else bool(r[i + 1]),
                }
                for label, i in (("t1", 6), ("t3", 8), ("t5", 10))
            },
            "filled_at": None if r[12] is None else str(r[12]),
        }
        for r in rows
    ]
    return {
        "count": len(items),
        "items": items,
        "note": "未回填的档位为 null(需要未来的 K 线才算得出来), 不用推算值填充; 命中 = 收益 > 0(平盘算未命中)。",
    }

File: frontend/packages/api/src/decisions.ts
/**
 * 决策账本 API 客户端(B6 前端 / 设计稿 v3.0 §八「新增能力」)。
 *
 * 后端契约(2026-09-18, v175 `decision_log`):
 *  · `GET /api/decisions/stats?days&min_sample` —— 按信号类型统计 T+1/3/5 命中率。
 *    **样本不足时 `hit_rate` 为 null 且 `insufficient=true`** —— 页面必须显示"样本不足",
 *    不许把 null 当 0 或自己算百分比(诚实口径)。
 *  · `GET /api/decisions/log?kind&limit` —— 信号明细(当时价格 + 证据快照 + 回填状态)。
 *    `ret/hit` 为 null 表示"还没回填"(需要未来的 K 线才算得出来), 不是 0。
 */
import { fetchAPI } from './client'

/** 单个时间档(T+1 / T+3 / T+5)的统计。 */
export interface DecisionHorizon {
  /** 已回填样本数(**只算有真实结果的行**, 未回填不进分母) */
  n: number
  /** 命中率; 样本不足或尚无样本时为 null */
  hit_rate: number | null
  /** true = 不给数字(页面显示"样本不足") */
  insufficient: boolean
  /** 为什么不给(空串=可以给) */
  note: string
}

export interface DecisionStatsRow {
  signal_kind: string
  n_total: number
  horizons: Record<'t1' | 't3' | 't5', DecisionHorizon>
}

export interface DecisionStatsResponse {
  since: string
  min_sample: number
  rows: DecisionStatsRow[]
  note: string
}

/** 一条信号明细。 */
export interface DecisionLogItem {
  signal_kind: string
  symbol: string
  trade_date: string
  /** 信号产生时的价格(元); **null = 当时没取到价, 不是 0** */
  price_at_signal: number | null
  /** 当时证据快照(JSON 字符串, 便于回放"凭什么出这个信号") */
  context: string
  source: string
  outcomes: Record<'t1' | 't3' | 't5', { ret: number | null; hit: boolean | null }>
  /** 回填时间; null = 尚未回填 */
  filled_at: string | null
}

export interface DecisionLogResponse {
  count: number
  items: DecisionLogItem[]
  note: string
}

/** 命中率统计(样本不足时不返回数字, 由调用方如实展示)。 */
export function fetchDecisionStats(days = 180, minSample = 30) {
  return fetchAPI<DecisionStatsResponse>(
    `/decisions/stats?days=${days}&min_sample=${minSample}`,
    { cacheMode: 'reload' },
  )
}

/** 信号明细(kind 省略 = 全部类型)。 */
export function fetchDecisionLog(kind?: string, limit = 100) {
  const q = new URLSearchParams({ limit: String(limit) })
  if (kind) q.set('kind', kind)
  return fetchAPI<DecisionLogResponse>(`/decisions/log?${q.toString()}`, { cacheMode: 'reload' })
}

/** 信号类型的中文名(后端用机器名, 页面给人看)。 */

export const SIGNAL_KIND_LABEL: Record<string, string> = {
  resonance3: '三指标共振',
  gs_buy: 'GS 买点',
  gs_sell: 'GS 卖点',
  auction_pool: '竞价池',
}

export function kindLabel(kind: string): string {
  return SIGNAL_KIND_LABEL[kind] ?? kind
}

/** 对象式导出(与本包既有风格一致: `xxxApi.get(...)`) */
export const decisionsApi = {
  stats: (days = 180, minSample = 30) => fetchDecisionStats(days, minSample),
  log: (kind?: string, limit = 100) => fetchDecisionLog(kind, limit),
}
